File: rplugin/python3/deoplete/sources/eskk.py
#=============================================================================
# FILE: eskk.py
# AUTHOR: Shougo Matsushita <Shougo.Matsu at gmail.com>
# License: MIT license  {{{
#     Permission is hereby granted, free of charge, to any person obtaining
#     a copy of this software and associated documentation files (the
#     "Software"), to deal in the Software without restriction, including
#     without limitation the rights to use, copy, modify, merge, publish,
#     distribute, sublicense, and/or sell copies of the Software, and to
#     permit persons to whom the Software is furnished to do so, subject to
#     the following conditions:
#
#     The above copyright notice and this permission notice shall be included
#     in all copies or substantial portions of the Software.
#
#     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
#     OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
#     MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
#     IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
#     CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
#     TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
#     SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
# }}}
#=============================================================================

import re
import deoplete.util

from .base import Base

class Source(Base):
    def __init__(self, vim):
        Base.__init__(self, vim)

        self.name = 'eskk'
        self.mark = '[eskk]'
        self.is_bytepos = True
        self.is_volatile = True
        self.is_skip_langmap = False
        self.rank = 500
        self.matchers = []
        self.sorters = []
        self.min_pattern_length = 0

    def get_complete_position(self, context):
        if not self.vim.call('eskk#is_enabled') or re.search(
                r'[a-zA-Z]$', context['input']):
            return -1
        return self.vim.call('eskk#complete#eskkcomplete', 1, '')

    def gather_candidates(self, context):
        return self.vim.call(
            'eskk#complete#eskkcomplete', 0, context['complete_str'])

File: denops/@ddc-sources/eskk.ts
import {
  BaseSource,
  Candidate,
  Context,
} from "https://deno.land/x/ddc_vim@v0.4.0/types.ts#^";
import { Denops } from "https://deno.land/x/ddc_vim@v0.4.0/deps.ts#^";

export class Source extends BaseSource {
  isBytePos = true;

  async getCompletePosition(args: {
    denops: Denops,
    context: Context,
  }): Promise<number> {
    const enabled = await args.denops.call('eskk#is_enabled') as boolean;
    if (!enabled) {
      return -1;
    }
    if (/[a-zA-Z]$/.test(args.context.input)) {
      return -1;
    }
    return args.denops.call('eskk#complete#eskkcomplete', 1, '');
  }

  async gatherCandidates(args: {
    denops: Denops,
    completeStr: string,
  }): Promise<Candidate[]> {
    return args.denops.call('eskk#complete#eskkcomplete', 0, args.completeStr);
  }
}
